File: serial_listener.php
<?php
require "php_serial.class.php";

$serial = new phpSerial();
$serial->deviceSet("/dev/ttyUSB0"); // Adjust for your serial device
$serial->confBaudRate(9600);
$serial->confParity("none");
$serial->confCharacterLength(8);
$serial->confStopBits(1);
$serial->deviceOpen();

$database = new mysqli("localhost", "user", "password", "home_security");

while (true) {
    $input = $serial->readPort();

    if ($input == "V") {
        shell_exec("raspivid -o /var/www/html/live.h264 -t 0 &");
        
        // Notify web server
        $ch = curl_init("http://yourserver.com/notify.php");
        curl_setopt($ch, CURLOPT_POST, 1);
        curl_setopt($ch, CURLOPT_POSTFIELDS, "event=recording_started");
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_exec($ch);
        curl_close($ch);
    } elseif ($input == "R") {
        $uid = "";
        for ($i = 0; $i < 8; $i++) {
            $uid .= $serial->readPort();
        }

        // Check UID in database
        $query = $database->prepare("SELECT pin FROM users WHERE uid = ?");
        $query->bind_param("s", $uid);
        $query->execute();
        $query->store_result();

        if ($query->num_rows > 0) {
            $query->bind_result($pin);
            $query->fetch();
            $serial->sendMessage("P" . $pin);
        } else {
            $serial->sendMessage("X");
        }
        $query->close();
    } elseif ($input == "W") {
        shell_exec("pkill raspivid"); // Stop recording
    }

    usleep(500000); // Wait half a second before checking again
}

$serial->deviceClose();
$database->close();
?>

File: serial_listener.py
import serial
import time
import subprocess
import requests
import mysql.connector

# Setup serial connection
ser = serial.Serial('/dev/ttyUSB0', baudrate=2400, timeout=1)

# Setup database connection
conn = mysql.connector.connect(
    host="localhost",
    user="your_db_user",
    password="your_db_password",
    database="your_db_name"
)
cursor = conn.cursor()

# Helper: send message to serial
def send_serial(message):
    ser.write(message.encode())

# Helper: start raspivid
def start_recording():
    subprocess.Popen(["raspivid", "-o", "/var/www/html/live.h264", "-t", "0"])
    requests.post("http://10.0.0.184:5555/notify.php", data={"event": "recording_started"})

# Helper: stop raspivid
def stop_recording():
    subprocess.call(["pkill", "-f", "raspivid"])

# Main loop
try:
    while True:
        input_char = ser.read().decode()

        if input_char == "V":
            start_recording()

        elif input_char == "R":
            uid = ser.read(8).decode()
            cursor.execute("SELECT pin FROM users WHERE uid = %s", (uid,))
            result = cursor.fetchone()

            if result:
                pin = result[0]
                send_serial("P" + pin)
            else:
                send_serial("X")

        elif input_char == "W":
            stop_recording()

        time.sleep(0.5)

except KeyboardInterrupt:
    print("Stopping script...")

finally:
    ser.close()
    cursor.close()
    conn.close()